### PR TITLE
Map behaviour tweaks

### PR DESCRIPTION
**Pull request by daviddrysdale**

### CODE REVIEW

File: server/map_rpc_server.go [hunk 1]
```diff
@@ -171,6 +169,14 @@ func (t *TrillianMapServer) getLeavesByRevision(ctx context.Context, mapID int64
 	}, nil
 }
 
+func checkIndexSize(index []byte, hasher hashers.MapHasher) error {
+	if len(index) != hasher.Size() {
+		// TODO(drysdale): why does the leaf index need to be the same size as the tree hash?
```
Comment (Martin2112) [inline server/map_rpc_server.go]: Drive-by: It's not an index, the API is misnamed. I just merged a PR that adds more doc to the proto.

File: storage/mysql/map_storage.go [hunk 2]
```diff
@@ -111,6 +111,10 @@ func (m *mySQLMapStorage) begin(ctx context.Context, tree *trillian.Tree) (stora
 		readRevision: -1,
 	}
 
+	if readonly {
+		return mtx, nil
```
Comment (AlCutter) [inline storage/mysql/map_storage.go]: Maybe add a comment that `readRevision` will be set by a call to `LatestSignedMapRoot`?

Comment (AlCutter) [reply storage/mysql/map_storage.go]: _might_ be worth making `LatestSignedMapRoot` be a lazy caching getter too?

File: server/map_rpc_server.go (hunk 1 above)
Comment (AlCutter) [reply server/map_rpc_server.go]: It is also an index, it's a large large number, but the hashed key value which gives the path to the leaf by definition is also the ordinal number of that leaf.

Strictly, the leaf index doesn't need to be the same size at the tree root hash (e.g. if the index was calculated using SHA512, but the tree hashes were SHA256), that's just how it is currently.

File: trillian_map_api.proto [hunk 3]
```diff
@@ -108,6 +108,11 @@ service TrillianMap {
   // For indexes that do not exist, the inclusion proof will use nil for the empty leaf value.
   rpc GetLeaves(GetMapLeavesRequest) returns(GetMapLeavesResponse) {}
   rpc GetLeavesByRevision(GetMapLeavesByRevisionRequest) returns(GetMapLeavesResponse) {}
+  // SetLeaves sets the values for the provided leaves, and returns the new map root if successful.
```
Comment (AlCutter) [inline trillian_map_api.proto]: Currently, I think it's tighter than that - the identical set of leaves must be retried. 
(Because partial writes of subtrees at rev n+1 are never deleted; if you set a superset of leaves then some of these subtrees [particularly up at the top of the tree] will need to be updated and there's no support for that right now.)

File: server/map_rpc_server.go (hunk 1 above)
Comment (Martin2112) [reply server/map_rpc_server.go]: Yes - true but people who come to this from the log code might expect an 'index' API to behave similarly, which isn't really the case.

File: storage/mysql/map_storage.go (hunk 2 above)
Comment (daviddrysdale) [reply storage/mysql/map_storage.go]: Added the comment; at the moment, lazily getting the map root wouldn't help anything.

File: trillian_map_api.proto (hunk 3 above)
Comment (daviddrysdale) [reply trillian_map_api.proto]: Made the text tighter.

File: server/map_rpc_server.go (hunk 1 above)
Comment (daviddrysdale) [reply server/map_rpc_server.go]: I agree with @Martin2112 that `index` is an unhelpful name, but it's at least a name that's consistent with all the current code...

Comment (Martin2112) [reply server/map_rpc_server.go]: I think it's too late to change it now.

Comment (gdbelvin) [reply server/map_rpc_server.go]: The key requirement is that all leaves in the map have the same `index` length.

File: storage/mysql/map_storage.go [hunk 4]
```diff
@@ -106,32 +106,41 @@ func (m *mySQLMapStorage) begin(ctx context.Context, tree *trillian.Tree) (stora
...
 	}
 
-	mtx.root, err = mtx.LatestSignedMapRoot(ctx)
+	if readonly {
+		// readRevision will be set later, by the first
+		// GetSignedMapRoot/LatestSignedMapRoot operation.
+		return mtx, nil
+	}
+
+	root, err := mtx.LatestSignedMapRoot(ctx)
```
Comment (gdbelvin) [inline storage/mysql/map_storage.go]: Should the rest of this function be broken into a separate function that `GetSignedMapRoot` can also call?

File: storage/mysql/map_storage_test.go [hunk 5]
```diff
@@ -99,20 +99,23 @@ func TestMapSnapshot(t *testing.T) {
...
-				t.Fatalf("%v: err = %q, wantErr = %v", test.desc, err, test.wantErr)
-			} else if hasErr {
-				return
+			if err != nil {
+				t.Fatalf("SnapshotForTree()=_,%v; want _, nil", err)
 			}
 			defer tx.Close()
 
 			_, err = tx.LatestSignedMapRoot(ctx)
 			if err != nil {
```
Comment (gdbelvin) [inline storage/mysql/map_storage_test.go]: This logic would be a bit simpler if we used `got` / `want`

Comment (daviddrysdale) [reply storage/mysql/map_storage_test.go]: Fixed.

File: storage/mysql/map_storage.go (hunk 4 above)
Comment (daviddrysdale) [reply storage/mysql/map_storage.go]: I don't think so -- both versions of `GetSignedMapRoot` (mapserver and storage) deal with `trillian.SignedMapRoot` objects rather than `types.MapRootV1` objects

**Comment by daviddrysdale** [comment]
Added different text to `checkIndexSize()` (in server/map_rpc_server.go) which hopefully encapsulates the discussion.
